src/Dist_Archive_Command.php

```diff
@@ -224,44 +192,74 @@ public function __invoke( $args, $assoc_args ) {
 				$archive_filename .= '.tar.gz';
 			}
 		}
-		$archive_filepath = $archive_path . '/' . $archive_filename;
+		$archive_absolute_filepath = $archive_path . '/' . $archive_filename;
 
 		chdir( dirname( $source_path ) );
 
 		if ( Utils\get_flag_value( $assoc_args, 'create-target-dir' ) ) {
-			$this->maybe_create_directory( $archive_filepath );
+			$this->maybe_create_directory( $archive_absolute_filepath );
 		}
 
 		if ( ! is_dir( dirname( $archive_path ) ) ) {
 			WP_CLI::error( "Target directory does not exist: {$archive_path}" );
 		}
 
-		if ( 'zip' === $assoc_args['format'] ) {
-			$excludes = implode( ' --exclude ', $ignored_files );
-			if ( ! empty( $excludes ) ) {
-				$excludes = ' --exclude ' . $excludes;
+		// If the files are being zipped in place, we need the exclusion rules.
+		// whereas if they were copied for any reasons above, the rules have already been applied.
+		if ( $source_path !== $path || empty( $file_ignore_rules ) ) {
+			if ( 'zip' === $assoc_args['format'] ) {
+				$cmd = "zip -r '{$archive_absolute_filepath}' {$archive_base}";
+			} elseif ( 'targz' === $assoc_args['format'] ) {
+				$cmd = "tar -zcvf {$archive_absolute_filepath} {$archive_base}";
+			}
+		} else {
+			$tmp_dir = sys_get_temp_dir() . DIRECTORY_SEPARATOR . uniqid( $archive_base . '.' . $version );
+			mkdir( $tmp_dir, 0777, true );
+			if ( 'zip' === $assoc_args['format'] ) {
+				$include_list_filepath = $tmp_dir . '/include-file-list.txt';
+				file_put_contents(
+					$include_list_filepath,
+					trim(
+						implode(
+							"\n",
+							array_map(
+								function ( $relative_path ) use ( $source_path ) {
+									return basename( $source_path ) . $relative_path;
+								},
+								$this->get_file_list( $source_path )
+							)
+						)
+					)
+				);
+				$cmd = "zip -r '{$archive_absolute_filepath}' {$archive_base} -i@{$include_list_filepath}";
+			} elseif ( 'targz' === $assoc_args['format'] ) {
+				$exclude_list_filepath = $tmp_dir . '/exclude-file-list.txt';
+				$excludes              = array_filter(
+					array_map(
+						function ( $ignored_file ) use ( $source_path ) {
+							if ( php_uname( 's' ) === 'Linux' ) {
+								return preg_quote( basename( $source_path ) . $ignored_file, '\\' );
+							} else {
+								return '^' . preg_quote( basename( $source_path ) . $ignored_file, '\\' ) . '$';
+							}
```

Comment:
I think the intention and maintainability is clearer here if this is split up into a different sequence:
```suggestion
							$regex = preg_quote( basename( $source_path ) . $ignored_file, '\\' );
							return ( php_uname( 's' ) === 'Linux' ) ? $regex : "^{$regex}$";
```